### ISSUE TITLE
Unique constraint violation on name key in snis table

### ISSUE THREAD

**Issue by rgershkovich** [Kong/kubernetes-ingress-controller#676]
# Summary
After upgrading to KIC 0.8.1 and Kong 2.0.4 from KIC 0.6.2 and Kong 1.3 we starting getting such errors in one of our environments:

```console
I0521 19:37:38.932343       1 controller.go:120] syncing Ingress configuration...
I0521 19:37:38.932758       1 parser.go:1655] getting endpoints for service ns/svc_name and port &ServicePort{Name:svc_name,Protocol:TCP,Port:80,TargetPort:{1 0 svc_name},NodePort:0,}
I0521 19:37:38.932797       1 parser.go:1698] endpoints found: [{10.97.12.16 80}]
...
W0521 19:37:38.933340       1 parser.go:368] Deprecated KongCredential in use, please use secret-based credentials. KongCredential resource will be removed in future.
E0521 19:39:25.797595       1 controller.go:127] unexpected failure updating Kong configuration:
10 errors occurred:
	while processing event: {Create} failed: 409 Conflict {"message":"UNIQUE violation detected on '{name=\"xxx\"}'","name":"unique constraint violation","fields":{"name":"xxx"},"code":5}
	while processing event: {Create} failed: 409 Conflict {"message":"UNIQUE violation detected on '{name=\"yyy\"}'","name":"unique constraint violation","fields":{"name":"yyy"},"code":5}
...
```

At the same time there are corresponding errors in lua logs:

```console
2020/05/21 18:47:43 [notice] 79#0: *147118 [lua] init.lua:394: upsert(): ERROR: duplicate key value violates unique constraint "snis_name_key"
Key (name)=(xxx) already exists., client: 127.0.0.1, server: kong_admin, request: "PUT /snis/75cf49b1-e2ed-4d18-aff7-25807e8615e9 HTTP/1.1", host: "127.0.0.1:8444"
```

**Kong Ingress controller version**
0.8.1

**Kong or Kong Enterprise version**
Standard Kong 2.0.4

Tried various combinations – same behaviour with 0.8.0/0.8.1+2.0.x releases.

### Kubernetes version
```console
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-13T11:52:32Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15+", GitVersion:"v1.15.9-gke.24", GitCommit:"39e41a8d6b7221b901a95d3af358dea6994b4a40", GitTreeState:"clean", BuildDate:"2020-02-29T01:24:35Z", GoVersion:"go1.12.12b4", Compiler:"gc", Platform:"linux/amd64"}
```

### Environment
- GKE
- cos
```
BUILD_ID=12371.175.0
NAME="Container-Optimized OS"
KERNEL_COMMIT_ID=1be45a63ab45e4483c7f5aa32a0d0390e8b5a9b9
GOOGLE_CRASH_ID=Lakitu
VERSION_ID=77
BUG_REPORT_URL="https://cloud.google.com/container-optimized-os/docs/resources/support-policy#contact_us"
PRETTY_NAME="Container-Optimized OS from Google"
VERSION=77
GOOGLE_METRICS_PRODUCT_ID=26
HOME_URL="https://cloud.google.com/container-optimized-os/docs"
ID=cos
```
- 4.19.102+

- **Install tools**: 
https://raw.githubusercontent.com/Kong/kubernetes-ingress-controller/master/deploy/single/all-in-one-postgres.yaml
, but the PgSQL DB is PostgreSQL 9.6 CloudSQL (google managed).

### What happened
I see that the snis table in PgSQL is populated and the number of entries there is the same as the number of certificates. However the ID which is shown in PUT request is different from any id found in the table.

The weird thing is that we have 4 environments and this happened only in a single one.. I struggle to understand what is so special about it and how to provide more logs.

Before upgrading I specifically checked that there were 0 issues in logs of 0.6.2 version. Of course no new resources were added or changed in cluster besides Kong components themselves during upgrade. Also when downgrading to 0.7.1 these errors go away (image version is the only thing changed in this experiment). Seeing that this happens in `syncIngress` function I tried to check for any potential collisions between Ingress resources but it looks like changes I make to them don't propagate anyway.

### Expected behavior
No errors about "snis_name_key" unique constraint violations.

### Steps To Reproduce
Unfortunately, I am not sure how to reproduce the issue exactly, as we have a relatively large number of resources in the cluster..

**Comment by hbagdi** [comment]
Some questions to better understand this.
Is the sni `xxx` or `yyy` present in more than 1 Ingress resource? The `sni` here is `host` inside the `tls` section of an Ingress resource. Also, are these SNIs associated with different certificates at the same time by any chance?

**Comment by rgershkovich** [comment]
Thank you for the quick reply and for you work on the controller!

> Is the sni `xxx` or `yyy` present in more than 1 Ingress resource
> The sni here is `host` inside the `tls` section of an Ingress resource

No, we do not have overlapping hosts in different Ingress resources. Double-checked with
`kubectl get ingress -A -o json | jq -r '.items[].spec.tls[].hosts[]' | sort | uniq -c`,
got 1 for every record.

> Also, are these SNIs associated with different certificates at the same time by any chance

Not to my knowledge.. We have `cert-manager` in the cluster, all of our certs are LE certs issued by `cert-manager`, and all Ingress resources have only 1 `host` with an annotation to use an http acme solver, so 1 Ingress should be associated with 1 Certificate and normally 1 KongIngress (we use it to do https redirection, for example).

**Comment by hbagdi** [comment]
> Before upgrading I specifically checked that there were 0 issues in logs of 0.6.2 version. Of course no new resources were added or changed in cluster besides Kong components themselves during upgrade

On re-reading, I think that this is an issue in decK(the underlying library doing the syncing).
And I think this is an issue with upgrade only, meaning, if you fix it once, then it shouldn't happen again.

I do want to get to root of this as it could be a potential bug in another subsystem.


> 2020/05/21 18:47:43 [notice] 79#0: *147118 [lua] init.lua:394: upsert(): ERROR: duplicate key value violates unique constraint "snis_name_key"
Key (name)=(xxx) already exists., client: 127.0.0.1, server: kong_admin, request: "PUT /snis/75cf49b1-e2ed-4d18-aff7-25807e8615e9 HTTP/1.1", host: "127.0.0.1:8444"

What this says is that `xxx` SNI is associated with another ID in the table.
Is the ID of `xxx` SNI in the table different than `75cf49b1-e2ed-4d18-aff7-25807e8615e9`?

**Comment by rgershkovich** [comment]
When I `select * from snis`, I do not see ID's in lua logs resulting in 409's match any of DB records, neither `id` nor `certificate_id`.

---

I tried to experiment more and here is what I've found. In our dev environment I left Kong controller with only one single host and downgraded this setup to 0.6.2 and tried upgrading to 0.8.1 once again.

It turned out that the same problem appeared, but it resolved on its own.
Several attempts with different ID's every time shown in logs:
```
2020/05/22 08:39:06 [notice] 22#0: *641 [lua] init.lua:394: upsert(): ERROR: duplicate key value violates unique constraint "snis_name_key"
Key (name)=(dev_fqdn) already exists., client: 127.0.0.1, server: kong_admin, request: "PUT /snis/84c41cc2-7661-4486-8c22-a6bdfebe7030 HTTP/1.1", host: "127.0.0.1:8444"
127.0.0.1 - - [22/May/2020:08:39:06 +0000] "PUT /snis/84c41cc2-7661-4486-8c22-a6bdfebe7030 HTTP/1.1" 409 183 "-" "Go-http-client/1.1"
...
2020/05/22 08:39:10 [notice] 22#0: *701 [lua] init.lua:394: upsert(): ERROR: duplicate key value violates unique constraint "snis_name_key"
Key (name)=(dev_fqdn) already exists., client: 127.0.0.1, server: kong_admin, request: "PUT /snis/30ef9e48-da5b-4604-8747-29954f8b5823 HTTP/1.1", host: "127.0.0.1:8444"
127.0.0.1 - - [22/May/2020:08:39:10 +0000] "PUT /snis/30ef9e48-da5b-4604-8747-29954f8b5823 HTTP/1.1" 409 183 "-" "Go-http-client/1.1"
...
2020/05/22 08:39:13 [notice] 22#0: *701 [lua] init.lua:394: upsert(): ERROR: duplicate key value violates unique constraint "snis_name_key"
Key (name)=(dev_fqdn) already exists., client: 127.0.0.1, server: kong_admin, request: "PUT /snis/7ac8b828-ef5f-4fa2-b09a-408db548f4cd HTTP/1.1", host: "127.0.0.1:8444"
127.0.0.1 - - [22/May/2020:08:39:13 +0000] "PUT /snis/7ac8b828-ef5f-4fa2-b09a-408db548f4cd HTTP/1.1" 409 183 "-" "Go-http-client/1.1"
```

The id in `snis` table is `489923f3-568d-4e2a-8e40-230dc5a80717`.
Then in 30 seconds it made a request with the id equal to DB record:

```console
127.0.0.1 - - [22/May/2020:08:39:16 +0000] "PUT /snis/489923f3-568d-4e2a-8e40-230dc5a80717 HTTP/1.1" 200 210 "-" "Go-http-client/1.1"
```

At the same time in controller logs:
```console
E0522 08:39:13.366264       1 controller.go:127] unexpected failure updating Kong configuration:
1 errors occurred:
	while processing event: {Create} failed: 409 Conflict {"message":"UNIQUE violation detected on '{name=\"dev_fqdn\"}'","name":"unique constraint violation","fields":{"name":"dev_fqdn"},"code":5}
W0522 08:39:13.366332       1 queue.go:112] requeuing kong/kong-validation-webhook, err 1 errors occurred:
	while processing event: {Create} failed: 409 Conflict {"message":"UNIQUE violation detected on '{name=\"dev_fqdn\"}'","name":"unique constraint violation","fields":{"name":"dev_fqdn"},"code":5}
I0522 08:39:16.660290       1 kong.go:71] successfully synced configuration to Kong
```

If that can be of any help, these are migrations applied during this process:
```console
New migrations available:
      core: 006_130_to_140, 007_140_to_150, 008_150_to_200
 hmac-auth: 002_130_to_140
    oauth2: 003_130_to_140
       jwt: 002_130_to_140
basic-auth: 002_130_to_140
  key-auth: 002_130_to_140
       acl: 002_130_to_140
      acme: 000_base_acme
```

Sorry if this is a result of misconfiguration from our side.. Let me know if/how I can gather any additional debug info.

**Comment by hbagdi** [comment]
I think this is related to this change: https://github.com/hbagdi/deck/commit/acdb161954b3081b0e38cb827cb3b685bf96b072#diff-4ac32a78649ca5bdd8e0ba38b7006a1eR34 which breaks the upgrade path but that should either resolve itself in most cases or require manual intervention in some cases.

Do you see this happening again once it fixes itself?

**Comment by rgershkovich** [comment]
Thank you for getting back.

The issue is that in one of our environments this has not resolved automatically in at least 8 hours. Because further updates to config were blocked by this, we had to roll back to previous versions.

Could you please recommend a workaround for this/how to manually fix?

**Comment by hbagdi** [comment]
> Could you please recommend a workaround for this/how to manually fix?

If the following is feasible, it should fix the problem for you:
1. Stop the controller (old version)
2. Drop all the rows in certificates and snis table in Kong (either psql cli or simply use Kong's Admin API)
3. Start the controller (new version)

That _should_ fix this problem.

**Comment by rgershkovich** [comment]
Did as you suggested – looks like no issues during upgrade, new version repopulated tables correctly, thank you for the help.

**Comment by hbagdi** [comment]
Closing this as the problem is resolved. Please re-open if needed.

**Comment by dwin** [comment]
Also encountered this issue. Removing all SNI table entries without removing certificates fixed the issue.

**Comment by hbagdi** [comment]
Please open a new issue if your controller version is 0.8.1. I was suppose to close this issue but hit the wrong button after my last comment.
